modifiers/prop_monitor.h

```diff
@@ -22,10 +22,34 @@
 
 namespace pono {
 
+/** Adds a monitor state variable to the transition system for the given
+ *  property. The monitor variable is initialized to true and holds the value of
+ * `prop` at the previous time step. The transition system is modified in place.
+ *
+ *  The function can only be used when
+ *  (1) `prop` contains a next-state variable (and `ts` is relational) or
+ *  (2) the transition relation `ts.trans()` is right-total.
+ *  This is to ensure that `ts` allows at least one further transition when
+ * `prop` is falsified (recall that the monitor has a one-step delay).
+ *
+ *  NOTE: only works for safety properties currently.
+ *
+ *  @param ts the transition system to modify
+ *  @param prop the property term to monitor
+ *  @return the monitor state variable term
+ */
 smt::Term add_prop_monitor(TransitionSystem & ts, const smt::Term & prop)
 {
+  // only in debug mode as right-total check can be expensive
+  assert(!ts.no_next(prop) || ts.is_right_total());
   logger.log(1, "Adding a monitor for the property");
 
+  // checks for functional TS
+  if (ts.is_functional() && !ts.no_next(prop)) {
+    throw PonoException(
+        "Cannot use next in property of a functional transition system.");
+  }
+
```

Comment:
What's the reason for moving this call here? Now this function will call `ts.no_next()` twice, which adds some overhead, given that it has to traverse the property twice.